javaslang/src/main/java/javaslang/collection/Arrays2.java

```diff
@@ -0,0 +1,92 @@
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.util.Random;
+
+/**
+ * Internal class, containing Java array manipulation helpers.
+ * Many arrays are represented as simple Objects, to avoid casts at the client site, and to work with all types of arrays, including primitive ones.
+ *
+ * @author Pap Lőrinc
+ * @since 3.0.0
+ */
+@SuppressWarnings({"unchecked", "SuspiciousArrayCast"})
+final class Arrays2 { // TODO reuse these in `Array` also
+    private static final Object[] EMPTY = {};
+
+    static <T> T[] emptyArray() { return (T[]) EMPTY; }
+    static boolean isNullOrEmpty(Object[] array) {
+        return (array == null) || (array.length == 0);
+    }
+
+    static <T> T getOrDefault(Object[] array, int i, T defaultValue) {
+        return (isNullOrEmpty(array) || (i < 0) || (i >= array.length))
+               ? defaultValue
+               : (T) getAt(array, i);
+    }
+
+    /** Create a single element array */
+    static Object[] asArray(Object element) {
+        final Object[] copy = arrayCopy(1, null, 0, 0, 0);
+        copy[0] = element;
+        return copy;
+    }
+
+    /** System.arrayCopy with same source and destination */
+    static Object[] copyRange(Object array, int from, int to) {
+        final int length = to - from;
+        return arrayCopy(length, array, from, 0, length);
+    }
+
+    static Object[] copyUpdate(Object[] array, int index, Object element) {
+        if (array == null) { array = emptyArray(); }
+        final int length = array.length;
+        final Object[] copy = arrayCopy(Math.max(length, index + 1), array, 0, 0, length);
+        copy[index] = element;
+        return copy;
+    }
+
+    static Object[] copyDrop(Object[] array, int index) {
+        final int length = array.length;
+        return arrayCopy(length, array, index, index, length - index);
+    }
+
+    static Object[] copyTake(Object[] array, int length) {
+        return (length <= 0) ? emptyArray()
+                             : copyRange(array, 0, length);
+    }
+
+    /** for performance reasons the array allocation and the System.arraycopy must be next to each other - even an if (size > 0) will make it slower */
+    static Object[] arrayCopy(int arraySize, Object source, int sourceFrom, int destinationFrom, int size) {
+        if (size == 0) {
+            return new Object[arraySize];
+        } else {
+            final Object[] result = new Object[arraySize];
+            System.arraycopy(source, sourceFrom, result, destinationFrom, size);
+            return result;
+        }
+    }
+
+    static <T> T getAt(Object arrayObject, int index) {
+        final Object[] array = (Object[]) arrayObject;
+        return (T) array[index];
+    }
+
+    /** Randomly mutate array positions */
+    static <T> int[] shuffle(int[] array, Random random) {
```

Comment:
Please move `shuffle` and `swap` out of this class to the benchmark package. These (mutating) methods are not needed in production code.